M2/Macaulay2/d/boost-regex.cpp

```diff
@@ -0,0 +1,290 @@
+  static std::map<ExpressionKey, boost::regex> cache;
+
+  /* attempt to recover a compiled expression from a map */
+  auto key = ExpressionKey {strings_hash(pattern), flags};
+  auto hit = cache.find(key);
+  if (hit != cache.end()) return hit->second;
+#if DEBUG_REGEX & 1
+  std::cerr << "flags:\t" << flags << std::endl
+            << "regexp:\t" << M2_tocharstar(pattern) << std::endl;
+#endif
+
+  /* parse and set the flags */
+  boost::regex_constants::syntax_option_type regex_flags =
+      boost::regex::no_except; /* never throw exceptions */
+  regex_flags |= flags & extended ? boost::regex::extended : 0;
+  regex_flags |= flags & literal ? boost::regex::literal : 0;
+  regex_flags |= flags & icase ? boost::regex::icase : 0;
+  regex_flags |= flags & nosubs ? boost::regex::nosubs : 0;
+  regex_flags |= flags & collate ? boost::regex::collate : 0;
+  regex_flags |= flags & no_mod_m ? boost::regex::no_mod_m : 0;
+  regex_flags |= flags & no_mod_s ? boost::regex::no_mod_s : 0;
+  /* the following are on by default on POSIX ERE, so we disable them */
+  regex_flags &= flags & no_escape_in_lists ? regex_flags
+                                            : ~boost::regex::no_escape_in_lists;
+  regex_flags &= flags & no_bk_refs ? regex_flags : ~boost::regex::no_bk_refs;
+
+  /* compile the state machine */
+  boost::regex expression(M2_tocharstar(pattern), regex_flags);
+  if (expression.status() != 0)
+    std::cerr << "regex: could not compile the regular expression: "
+              << M2_tocharstar(pattern) << std::endl;
+
+  /* cache the compiled state machine */
+  auto res = cache.emplace(key, expression);
+  if (!res.second)
+    std::cerr << "regex: cache conflict detected between compiled expressions: "
+              << res.first->second.str() << " (old expression) and "
+              << expression.str() << " (new expression)" << std::endl;
+
+  return expression;
+}
+
+extern "C" {
+
+/* returns an array of pairs (s, r), indicating starting point
+ * and length of the first match and its capture groups */
+M2_arrayint regex_search(const M2_string pattern,
+                         const size_t start,
+                         const size_t range,
+                         const M2_string text,
+                         long flags)
+{
+#if DEBUG_REGEX & 2
+  std::cerr << "string:\t" << M2_tocharstar(text) << std::endl;
+#endif
+
+  M2_arrayint m = M2_makearrayint(0);
+  if (text->len < start) return m;
+
+  /* setting the default flags for search */
+  flags = flags == -1 ? perl | no_mod_s : flags;
+
+  auto expression = regex_compile(pattern, flags);
+  if (expression.status() != 0) return m;
+
+  bool status = false;
+  boost::cmatch matches {};
+
+  auto head = (const char*)&text->array + start;
+  auto tail = (const char*)&text->array + text->len;
+
+  boost::regex_constants::match_flag_type flag = boost::match_default;
+  flag |= flags & match_any ? boost::match_any : flag;
```

Comment:
yes, `flag|flag` yields `flag`, but `flag|0` would be clearer.

Reply:
This wouldn't always compile, since `0` doesn't have the type `match_flag_type`.

Reply:
I considered this, but there's a type error (at least when using clang) when using the suggestion, so I'll mark this as resolved.